scene/3d/chain_ik_3d.cpp

```diff
@@ -0,0 +1,484 @@
+			}
+		} else if (what == "extend_end_bone") {
+			r_ret = is_end_bone_extended(which);
+		} else if (what == "joint_count") {
+			r_ret = get_joint_count(which);
+		} else if (what == "joints") {
+			int idx = path.get_slicec('/', 3).to_int();
+			String prop = path.get_slicec('/', 4);
+			if (prop == "bone_name") {
+				r_ret = get_joint_bone_name(which, idx);
+			} else if (prop == "bone") {
+				r_ret = get_joint_bone(which, idx);
+			} else {
+				return false;
+			}
+		} else {
+			return false;
+		}
+	}
+	return true;
+}
+
+void ChainIK3D::get_property_list(List<PropertyInfo> *p_list) const {
+	String enum_hint;
+	Skeleton3D *skeleton = get_skeleton();
+	if (skeleton) {
+		enum_hint = skeleton->get_concatenated_bone_names();
+	}
+
+	LocalVector<PropertyInfo> props;
+
+	for (uint32_t i = 0; i < settings.size(); i++) {
+		String path = "settings/" + itos(i) + "/";
+		props.push_back(PropertyInfo(Variant::STRING, path + "root_bone_name", PROPERTY_HINT_ENUM_SUGGESTION, enum_hint));
+		props.push_back(PropertyInfo(Variant::INT, path + "root_bone", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_NO_EDITOR));
+		props.push_back(PropertyInfo(Variant::STRING, path + "end_bone_name", PROPERTY_HINT_ENUM_SUGGESTION, enum_hint));
+		props.push_back(PropertyInfo(Variant::INT, path + "end_bone", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_NO_EDITOR));
+		props.push_back(PropertyInfo(Variant::BOOL, path + "extend_end_bone"));
+		props.push_back(PropertyInfo(Variant::INT, path + "end_bone/direction", PROPERTY_HINT_ENUM, SkeletonModifier3D::get_hint_bone_direction()));
+		props.push_back(PropertyInfo(Variant::FLOAT, path + "end_bone/length", PROPERTY_HINT_RANGE, "0,1,0.001,or_greater,suffix:m"));
+		props.push_back(PropertyInfo(Variant::INT, path + "joint_count", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_ARRAY, "Joints," + path + "joints/,static,const"));
+		for (uint32_t j = 0; j < chain_settings[i]->joints.size(); j++) {
+			String joint_path = path + "joints/" + itos(j) + "/";
+			props.push_back(PropertyInfo(Variant::STRING, joint_path + "bone_name", PROPERTY_HINT_ENUM_SUGGESTION, enum_hint, PROPERTY_USAGE_EDITOR | PROPERTY_USAGE_READ_ONLY | PROPERTY_USAGE_STORAGE));
+			props.push_back(PropertyInfo(Variant::INT, joint_path + "bone", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_NO_EDITOR | PROPERTY_USAGE_READ_ONLY));
+		}
+	}
+
+	for (PropertyInfo &p : props) {
+		_validate_dynamic_prop(p);
+		p_list->push_back(p);
+	}
+}
+
+void ChainIK3D::_validate_dynamic_prop(PropertyInfo &p_property) const {
+	PackedStringArray split = p_property.name.split("/");
+	if (split.size() > 2 && split[0] == "settings") {
+		int which = split[1].to_int();
+
+		// Extended end bone option.
+		bool force_hide = false;
+		if (split[2] == "extend_end_bone" && get_end_bone(which) == -1) {
+			p_property.usage = PROPERTY_USAGE_NONE;
+			force_hide = true;
+		}
+		if (force_hide || (split[2] == "end_bone" && !is_end_bone_extended(which) && split.size() > 3)) {
```

Comment:
This will not hide `end_bone_name`
should `end_bone_name` should be inside `end_bone` (`end_bone/name`)?